docs: improve Zoe jsDoc annotations to support contract devs

Chris-Hibbert opened the pull request:
Improve jsDoc annotations in Zoe and ZoeHelpers to support contract devs.

import more types
declare zcf's type in the contracts
use OfferRecord (vs. Offer) where warranted
typedef for Invite that works.
typedef for opaque handles is {}. Typedefs are apparently aliases, so they're not distinct. typedef to Object with no properties is identical to `any`, which is less helpful.

JSDoc/Typedef doesn't seem to have a way of saying a record with an arbitrary number of typed Key-Value pairs
I describe one pair and append "|any".

If Invite is typedef Payment, then VSCode shows it as Payment everywhere. We typedef it as an Object with a getAllegedBrand, so it's a distinct type.

issue: #1007

Chris-Hibbert commented on packages/zoe/src/contracts/simpleExchange.js:
```diff
@@ -14,134 +20,135 @@ import { makeZoeHelpers, defaultAcceptanceMsg } from '../contractSupport';
-    };
-  }
+export const makeContract = harden(
+  /** @param {ContractFacet} zcf */ zcf => {
```
This comment and the changes above are the only non-white spaces changes in the file. The rest are driven by eslint.  This declaration allows vscode to typecheck arguments to all calls on zcf functions.

Chris-Hibbert commented on packages/zoe/src/contracts/multipoolAutoswap.js:
```diff
@@ -23,611 +30,624 @@ import {
-  // This contract must have a "central token" issuer in the terms.
-  const CENTRAL_TOKEN = 'CentralToken';
+export const makeContract = harden(
+  /** @param {ContractFacet} zcf */ zcf => {
```
This comment and the changes above are the only non-white spaces changes in the file. The rest are driven by eslint. This declaration allows vscode to typecheck arguments to all calls on zcf functions.

katelynsills commented on packages/zoe/src/contractSupport/zoeHelpers.js:
```diff
@@ -8,7 +8,7 @@ import { HandledPromise } from '@agoric/eventual-send';
  * @typedef {import('../zoe').Invite} Invite
  * @typedef {import('../zoe').OfferHook} OfferHook
  * @typedef {import('../zoe').CustomProperties} CustomProperties
- * @typedef {any} TODO Needs to be typed
```
I don't know the answer to this question, but do we need to retain this line? It looks like nothing changed in the file to change this line.

katelynsills commented on packages/zoe/src/contractSupport/zoeHelpers.js:
```diff
@@ -17,8 +17,14 @@ export const defaultAcceptanceMsg = `The offer has been accepted. Once the contr
+ *
+ * @param {ContractFacet} zcf
+ */
+// zcf only picks up the type if the param is in parens, which eslint dislikes
```
Huh, interesting. For what it is worth, prettier has moved to always using parens in 2.0.0 https://prettier.io/docs/en/options.html#arrow-function-parentheses

katelynsills commented on packages/zoe/src/zoe.js:
```diff
@@ -29,22 +29,36 @@ import { makeTables } from './state';
  * @typedef {any} TODO Needs to be typed
  * @typedef {string} Keyword
- * @typedef {Object} InstallationHandle
- * @typedef {Object.<Keyword,Issuer>} IssuerKeywordRecord
```
Was this {Object.<Keyword,Issuer>} notation not working?

katelynsills commented on packages/zoe/src/zoe.js:
```diff
@@ -147,36 +168,60 @@ import { makeTables } from './state';
+ * @property {OfferHandle} handle - opaque identifier for the offer, used as the key
+ * @property {InstanceHandle} instanceHandle - opaque identifier for the instance
+ * @property {Proposal} proposal - the offer proposal (including want, give, exit)
+ * @property {AmountKeywordRecord} amounts - the amountKeywordRecord that will be turned into payouts
```
```suggestion
 * @property {AmountKeywordRecord} currentAllocation - the amountKeywordRecord that will be turned into payouts
```

katelynsills replied:
Should we add notifier and updater here?

katelynsills replied:
Ok, that makes sense to me. Since I had two comments here, I just want to make sure the first suggested change comment doesn't get lost.

Chris-Hibbert replied:
I like the format of the built-in suggestions, but I haven't figured out how to acknowledge them when I'm making the change with a bunch of others on my side rather than hitting the github buttons.

I changed `amounts` to `currentAllocation`.

katelynsills commented on packages/zoe/src/zoe.js:
```diff
@@ -147,36 +168,60 @@ import { makeTables } from './state';
+ * @property {InstanceHandle} handle - opaque identifier for the instance, used as the table key
+ * @property {InstallationHandle} installationHandle - opaque identifier for the installation
+ * @property {Object.<string,function>} publicAPI - the invite-free publicly accessible API for the contract
+ * @property {object} terms - contract parameters
```
Should this be uppercase or lower case?

Chris-Hibbert replied:
The type system doesn't seem to care. Capitalized for consistency.

katelynsills commented on packages/zoe/src/zoe.js:
```diff
@@ -147,36 +168,60 @@ import { makeTables } from './state';
+ * @property {Object.<string,function>} publicAPI - the invite-free publicly accessible API for the contract
+ * @property {object} terms - contract parameters
+ * @property {IssuerKeywordRecord} issuerKeywordRecord - record with keywords keys, issuer values
+ * @property {KeywordRecord} keywords - an array of the keywords, in sorted order
```
This isn't a property

Chris-Hibbert replied:
done.

katelynsills commented on packages/zoe/src/zoe.js:
```diff
@@ -214,8 +259,8 @@ import { makeTables } from './state';
- * @param  {object[]} offerHandles - an array of offerHandles
- * @returns {TODO}
+ * @param  {OfferHandle[]} offerHandles - an array of offerHandles
+ * @returns {void}
```
Is this the right way to say 'undefined'?

Chris-Hibbert replied:
I couldn't find anything definitive. From the discussion I've seen, I'd prefer to omit it, but VSDoc interprets that as returning `any`, which is way wrong.

I'll use return `undefined`, as more consistent with JavaScript. With that VSCode shows `=> undefined` which isn't far from right.

katelynsills commented on packages/zoe/src/zoe.js:
```diff
@@ -230,7 +275,7 @@ import { makeTables } from './state';
  * offerHandle at the right time, and returns a contract-specific
  * OfferOutcome which will be put in the OfferResultRecord.
- * @param  {CustomProperties} customProperties - an object of
+ * @param  {CustomProperties} [customProperties] - an object of
```
Why the square brackets?

Chris-Hibbert replied:
It marks it as optional. (Displays the param with a `?`.)  I couldn't find a syntax using question mark directly.

michaelfig commented:
Please consider changing some of these types back.  Be forewarned that vscode sometimes gets confused so you have to do shift-cmd-P and select "restart ts language server" to see them properly.

Also, separating typedefs into their own block comments sometimes helps.  The TODOs are specifically to point out to documentors what needs to be done.

katelynsills commented:
@michaelfig what's the best place to learn about what the format of typedefs should be?

Chris-Hibbert commented:
> Please consider changing some of these types back. Be forewarned that vscode sometimes gets confused so you have to do shift-cmd-P and select "restart ts language server" to see them properly.

I'd be happy to change some back if it would improve something, but I don't understand what changes you would recommend. Which tool would you use to validate changes? 

I'd be happy to generalize if you can indicate which types of edits seem problematic. I described several particular changes and classes of changes in the top comment above. 

For each of my changes, I compared before and after in VSCode. When there was no change or no improvement, I reverted my changes, so all these had beneficial effects in VSCode.

> Also, separating typedefs into their own block comments sometimes helps. 

Yes, I found that to be the case, and did so in several cases that made types clearer when editing contracts. Is this a recommendation that I should do more of the same?

> The TODOs are specifically to point out to documentors what needs to be done.

I took them that way and updated several. (Invite, Offer, InstanceRecord, InstallationRecord) There were others that I left alone, either because I didn't find a better answer, or I wanted to limit the scope of the review and left some for a later pass. Is this a suggestion that it would have been better to leave more `TODO`s alone? Which ones? Was this in response to the removal of TODO in zoeHelpers.js? I didn't find any place where that hurt anything. Did I miss something?

michaelfig commented:
> JSDoc/Typedef doesn't seem to have a way of saying a record with an arbitrary number of typed Key-Value pairs

`@typedef {Object.<KeyType,ValueType>} FooRecord`

Or: `{Array<[K,V]>}` depending what you mean.

That's really the only regression I saw.  Your changes were good overall.

If it doesn't work, try breaking it out into a different block.

Cli checking is coming, but not this week.

michaelfig commented:
> @michaelfig what's the best place to learn about what the format of typedefs should be?

https://properjs.org has links

Chris-Hibbert commented:
>> JSDoc/Typedef doesn't seem to have a way of saying a record with an arbitrary number of typed Key-Value pairs

> @typedef {Object.<KeyType,ValueType>} FooRecord

I really wanted this to work. I tried several variants (including restarting the TS Server a few times), but VSCode keeps saying `any`.  But I'll admit my version isn't better. Since I couldn't do the repeaed fields, I fell back on `|any`, which has the same effect, so I'll revert it.

for the single pair version, this did actually work better for VSCode:
`@typedef {{[Keyword:string]:Amount}} Allocation`

> Or: {Array<[K,V]>} depending what you mean.

This works great for Arrays. Our FooKeywordRecords allow `{ Give: amount }`, `{ Buy: <amount>, Sell: <amount> }`, etc. I think if the `Object` typedef worked at all, it would only specify the first version.

> That's really the only regression I saw. Your changes were good overall.

Thanks.
